README.md
# feedyard/circleci-infra-agent [![CircleCI](https://circleci.com/gh/feedyard/circleci-infra-agent.svg?style=svg)](https://circleci.com/gh/feedyard/circleci-infra-agent) [![Docker Repository on Quay](https://quay.io/repository/feedyard/circleci-infra-agent/status "Docker Repository on Quay")](https://quay.io/repository/feedyard/circleci-infra-agent) [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/feedyard/circleci-infra-agent/master/LICENSE) [![Base Image](https://img.shields.io/badge/FROM-alpine-blue.svg)](https://alpinelinux.org)

Based on [feedyard/circleci-base-agent](https://github.com/feedyard/circleci-base-agent). includes common tools for buildiing and testing
aws or gcp infrastructure as code in circleci pipelines.

apk/bin               | gem            | pip
----------------------|----------------|----
terraform 0.12        | json           | boto3
packer                | awspec         | awscli
gcloud                | test-kitchen   | argparse
chamber               | kitchen-ec2    | jinja2
aws-iam-authenticator | kitchen-google |
tflint                |                |

See CHANGELOG for list of installed packages/versions. (version 4.4.0 and earlier for terraform 0.11)


CHANGELOG.md
# CHANGELOG

### 09-12-2019

FROM quay.io/feedyard/circleci-base-agent:4.12.0

JSON_VERSION=2.2.0
AWSPEC_VERSION=1.18.1
TEST_KITCHEN_VERSION=2.3.2
KITCHEN_EC2_VERSION=3.1.0
CHAMBER_VERSION=2.5.0
AWS_IAM_AUTHENTICATOR_VERSION=1.14.6
AWS_IAM_AUTHENTICATOR_RELEASE_DATE=2019-08-22
TFLINT_VERSION=0.11.1
boto3=1.9.228
awscli=1.16.238


### 06-20-2019 release version 4.7.0

tflint v0.8.3
awscli 1.16.183
boto3 1.9.173

### 06-19-2019 release version 4.6.0

feedyard/circleci-base-image:4.9.0
awscli 1.16.182
boto3 1.9.172

### 06-14-2019 terraform 12 update, add aws-iam-authenticator, package updates

TERRAFORM_VERSION=0.12.2
AWS_IAM_AUTHENTICATOR_VERSION=1.12.7
awscli 1.16.179
boto3 1.9.69

### 06-07-2019 new base image, package updates

feedyard/circleci-base-agent:4.5.0
awscli 1.16.164
boto3 1.9.164

### 05-18-2019 package updates

terraform 0.11.14
GCLOUS SDK 246.0.0
packer 1.4.1

### 04-05-2019 Publish version 4.4.0

Successfully installed
MarkupSafe-1.1.1
PyYAML-3.13
argparse-1.4.0
awscli-1.16.140
boto3-1.9.130
botocore-1.12.130
colorama-0.3.9
docutils-0.14
jinja2-2.10
jmespath-0.9.4
pyasn1-0.4.5
python-dateutil-2.8.0
rsa-3.4.2
s3transfer-0.2.0
urllib3-1.24.1

Successfully installed json-2.2.0
Successfully installed aws-partitions-1.149.0
Successfully installed aws-sdk-core-3.48.3
Successfully installed aws-sdk-acm-1.19.0
Successfully installed aws-sdk-acmpca-1.13.0
Successfully installed aws-sdk-apigateway-1.26.0
Successfully installed aws-sdk-alexaforbusiness-1.19.0
Successfully installed aws-sdk-amplify-1.3.0
Successfully installed aws-sdk-apigatewaymanagementapi-1.3.0
Successfully installed aws-sdk-apigatewayv2-1.4.0
Successfully installed aws-sdk-appmesh-1.6.0
Successfully installed aws-sdk-appstream-1.25.0
Successfully installed aws-sdk-appsync-1.12.0
Successfully installed aws-sdk-applicationautoscaling-1.22.0
Successfully installed aws-sdk-applicationdiscoveryservice-1.14.0
Successfully installed aws-sdk-athena-1.12.0
Successfully installed aws-sdk-autoscaling-1.20.0
Successfully installed aws-sdk-autoscalingplans-1.12.0
Successfully installed aws-sdk-backup-1.3.0
Successfully installed aws-sdk-batch-1.16.0
Successfully installed aws-sdk-budgets-1.18.0
Successfully installed aws-sdk-chime-1.6.0
Successfully installed aws-sdk-cloud9-1.11.0
Successfully installed aws-sdk-clouddirectory-1.14.0
Successfully installed aws-sdk-cloudformation-1.18.0
Successfully installed aws-sdk-cloudfront-1.15.0
Successfully installed aws-sdk-cloudhsm-1.12.0
Successfully installed aws-sdk-cloudhsmv2-1.12.0
Successfully installed aws-sdk-cloudsearch-1.9.0
Successfully installed aws-sdk-cloudsearchdomain-1.9.0
Successfully installed aws-sdk-cloudtrail-1.11.0
Successfully installed aws-sdk-cloudwatch-1.19.0
Successfully installed aws-sdk-cloudwatchevents-1.17.0
Successfully installed aws-sdk-cloudwatchlogs-1.17.0
Successfully installed aws-sdk-codebuild-1.32.0
Successfully installed aws-sdk-codecommit-1.17.0
Successfully installed aws-sdk-codedeploy-1.18.0
Successfully installed aws-sdk-codepipeline-1.15.0
Successfully installed aws-sdk-codestar-1.11.0
Successfully installed aws-sdk-cognitoidentity-1.10.0
Successfully installed aws-sdk-cognitoidentityprovider-1.16.0
Successfully installed aws-sdk-cognitosync-1.9.0
Successfully installed aws-sdk-comprehend-1.18.0
Successfully installed aws-sdk-comprehendmedical-1.3.0
Successfully installed aws-sdk-configservice-1.26.0
Successfully installed aws-sdk-connect-1.13.0
Successfully installed aws-sdk-costexplorer-1.21.0
Successfully installed aws-sdk-costandusagereportservice-1.10.0
Successfully installed aws-sdk-dax-1.11.0
Successfully installed aws-sdk-dlm-1.11.0
Successfully installed aws-sdk-datapipeline-1.9.0
Successfully installed aws-sdk-datasync-1.3.0
Successfully installed aws-sdk-databasemigrationservice-1.20.0
Successfully installed aws-sdk-devicefarm-1.19.0
Successfully installed aws-sdk-directconnect-1.16.0
Successfully installed aws-sdk-directoryservice-1.15.0
Successfully installed aws-sdk-docdb-1.4.0
Successfully installed aws-sdk-dynamodb-1.25.0
Successfully installed aws-sdk-dynamodbstreams-1.9.0
Successfully installed aws-sdk-ec2-1.78.0
Successfully installed aws-sdk-ecr-1.14.0
Successfully installed aws-sdk-ecs-1.36.0
Successfully installed aws-sdk-efs-1.13.0
Successfully installed aws-sdk-eks-1.15.0
Successfully installed aws-sdk-emr-1.14.0
Successfully installed aws-sdk-elasticache-1.14.0
Successfully installed aws-sdk-elasticbeanstalk-1.19.0
Successfully installed aws-sdk-elasticloadbalancing-1.12.0
Successfully installed aws-sdk-elasticloadbalancingv2-1.26.0
Successfully installed aws-sdk-elastictranscoder-1.11.0
Successfully installed aws-sdk-elasticsearchservice-1.19.0
Successfully installed aws-sdk-fms-1.12.0
Successfully installed aws-sdk-fsx-1.4.0
Successfully installed aws-sdk-firehose-1.14.0
Successfully installed aws-sdk-gamelift-1.15.0
Successfully installed aws-sdk-glacier-1.18.0
Successfully installed aws-sdk-globalaccelerator-1.4.0
Successfully installed aws-sdk-glue-1.30.0
Successfully installed aws-sdk-greengrass-1.17.0
Successfully installed aws-sdk-guardduty-1.14.0
Successfully installed aws-sdk-health-1.12.0
Successfully installed aws-sdk-iam-1.18.0
Successfully installed aws-sigv2-1.0.1
Successfully installed aws-sdk-importexport-1.9.0
Successfully installed aws-sdk-inspector-1.15.0
Successfully installed aws-sdk-iot-1.29.0
Successfully installed aws-sdk-iot1clickdevicesservice-1.11.0
Successfully installed aws-sdk-iot1clickprojects-1.10.0
Successfully installed aws-sdk-iotanalytics-1.16.0
Successfully installed aws-sdk-iotdataplane-1.9.0
Successfully installed aws-sdk-iotjobsdataplane-1.10.0
Successfully installed aws-sdk-kms-1.16.0
Successfully installed aws-sdk-kafka-1.3.0
Successfully installed aws-sdk-kinesis-1.13.1
Successfully installed aws-sdk-kinesisanalytics-1.12.0
Successfully installed aws-sdk-kinesisanalyticsv2-1.3.0
Successfully installed aws-sdk-kinesisvideo-1.12.0

profiles/cis-docker/controls/CIS-Docker-CE-1.1.0-Benchmark-4.rb
# frozen_string_literal: true
#
# inspec controls designed to run exec against docker://$ID target

title 'Container Images and Build File'

control 'cis-docker-benchmark-4.1' do
  impact 1.0
  title 'Ensure a user for the container has been created (Scored)'
  desc 'Create a non-root user for the container in the Dockerfile for the container image.'

  docker.containers.running?.ids.each do |id|
    describe docker.object(id) do
      skip 'ephemeral build container used on circleci - USER intentionally not defined'
      # its(%w(Config User)) { should_not eq nil }
    end
  end
end

control 'cis-docker-benchmark-4.2' do
  impact 1.0
  title 'Ensure that containers use trusted base images (Not Scored)'
  desc 'Ensure that the container image is written either from scratch or is based on another established and trusted base image downloaded over a secure channel.'

  ref url: 'https://alpinelinux.org'
  describe 'minimal base image test' do
    skip 'ogranizational decision to treat version 3.6 of official alpine image as trusted.'
  end
end

control 'cis-docker-benchmark-4.3' do
  impact 1.0
  title 'Ensure unnecessary packages are not installed in the container'
  desc 'Containers tend to be minimal and slim down versions of the Operating System. Do not install anything that does not justify the purpose of container.'

  ref url: 'https://alpinelinux.org'
  describe 'minimal base image test' do
    skip 'configuration controls validate base os family as alpine to prevent unknown packages in starting point. All other installs tested and intentional.'
  end
end

control 'cis-docker-benchmark-4.4' do
  impact 1.0
  title 'Ensure images are scanned and rebuilt to include security patches (Not Scored)'
  desc 'Images should be scanned frequently for any vulnerabilities. Rebuild the images to include patches and instantiate new containers from it.'

  describe 'scanned images and rebuild rather than patch' do
    skip 'organization also deploys images to quay.io which includes clair vulnerability scanning'
  end
end

control 'cis-docker-benchmark-4.5' do
  impact 1.0
  title 'Enable Content trust for Docker is enabled (Scored)'
  desc 'Content trust provides the ability to use digital signatures for data sent to and received from remote Docker registries. These signatures allow client-side verification of the integrity and publisher of specific image tags. This ensures provenance of container images. Content trust is disabled by default. You should enable it.'

  describe os_env('DOCKER_CONTENT_TRUST') do
    skip 'docker content trust not currently supported by quay.io'
    # its('content') { should eq '1' }
  end
end

control 'cis-docker-benchmark-4.6' do
  impact 0.0
  title 'Ensure HEALTHCHECK instruction have been added to the container image (Scored)'
  desc 'Add HEALTHCHECK instruction in your docker container images to perform the health check on running containers.'

  docker.containers.running?.ids.each do |id|
    describe docker.object(id) do
      skip 'conatiner has limited life span and circleci manages container orchestration'
      # its(%w(Config Healthcheck)) { should_not eq nil }
    end
  end
end

control 'cis-docker-benchmark-4.7' do
  impact 0.0
  title 'Ensure update instructions are not used alone in the Dockerfile (Not Scored)'
  desc 'Adding the update instructions in a single line on the Dockerfile will cache the update layer. Alternatively, you could use --no-cache flag during docker build process to avoid using cached layers.'

  # org requirement to include --no-cache option to apk add
  docker.images.ids.each do |id|
    describe command("docker history #{id}| grep -e '/add(?! --no-cache)/g'") do
      its('stdout') { should eq '' }
    end
  end
end

control 'cis-docker-benchmark-4.8' do
  impact 0.0
  title 'Ensure setuid and setgid permissions are removed in the images (Not Scored)'
  desc 'setuid and setgid permissions could be used for elevating privileges. Allow setuid and setgid permissions only on executables which need them. '

  ref url: 'http://man7.org/linux/man-pages/man2/setuid.2.html'
  ref url: 'http://man7.org/linux/man-pages/man2/setgid.2.html'
  # circleci agents are short-lived and this remediates the elevated privilege risk from attack
  # the following line could be used to alter executables with set permissions
  # RUN find / -perm +6000 -type f -exec chmod a-s {} \; || true
  describe 'setuid and setgid permissions' do
    skip 'starting from alpine base, all installed packages with executables use default'
  end
end

control 'cis-docker-benchmark-4.9' do
  impact 0.3
  title 'Ensure COPY is used instead of ADD in Dockerfile (Not Scored)'
  desc 'COPY instruction just copies the files from the local host machine to the container file system. ADD instruction potentially could retrieve files from remote URLs and perform operations such as unpacking.'

  docker.images.ids.each do |id|
    describe command("docker history #{id}| grep '/ADD(?! file:)/g'") do
      its('stdout') { should eq '' }
    end
  end
end

control 'cis-docker-benchmark-4.10' do
  impact 0.0
  title 'Ensure secrets are not installed in Dockerfiles (Not Scored)'
  desc 'Do not store any secrets in Dockerfile as these can be easily revealed using docker native commands.'

  describe 'Dockerfile test' do
    skip 'Manually verify that you have not used secrets in images and use tools such as Talisman to validate repo contents'
  end
end

control 'cis-docker-benchmark-4.11' do
  impact 0.0
  title 'Ensure verified packages are only installed (Not Scored)'
  desc 'Verifying authenticity of the packages is essential for building a secure container image.'

  describe 'Verify packages' do
    skip 'Use GPG keys for downloading and verifying packages or any other secure package distribution mechanism of your choice.'
  end
end

profiles/circleci-infra-agent/controls/circleci-infra-agent.rb
control 'python packages' do
  impact 1.0
  title 'confirm python package installation'
  desc 'confirm all desired python packages are installed'
  describe command('pip list') do
    its('stdout') { should include ('boto3') }
    its('stdout') { should include ('awscli') }
    its('stdout') { should include ('Jinja2') }
  end
end

control 'json' do
  impact 1.0
  title 'confirm json gem installed'
  desc 'confirm version of json installed'
  describe gem('json') do
    it { should be_installed }
    its('version') { should eq '2.2.0' }
  end
end

control 'awspec' do
  impact 1.0
  title 'confirm awspec gem installed'
  desc 'confirm version of awspec installed'
  describe gem('awspec') do
    it { should be_installed }
    its('version') { should eq '1.18.1' }
  end
end

control 'test-kitchen' do
  impact 1.0
  title 'confirm test-kitchen gem installed'
  desc 'confirm version of test-kitchen installed'
  describe gem('test-kitchen') do
    it { should be_installed }
    its('version') { should eq '2.3.2' }
  end
end

control 'kitchen-ec2' do
  impact 1.0
  title 'confirm kitchen-ec2 gem installed'
  desc 'confirm version of kitchen-ec2 installed'
  describe gem('kitchen-ec2') do
    it { should be_installed }
    its('version') { should eq '3.1.0' }
  end
end

control 'kitchen-google' do
  impact 1.0
  title 'confirm kitchen-google gem installed'
  desc 'confirm version of kitchen-google installed'
  describe gem('kitchen-google') do
    it { should be_installed }
    its('version') { should eq '2.0.1' }
  end
end

control 'aws-cli version' do
  impact 1.0
  title 'confirm aws-cli version installed'
  desc 'confirm version reported by aws-cli matches the desired version'
  describe command('aws --version') do
    its('stdout') { should include ('1.16') }
  end
end

# control 'gcloud version' do
#   impact 1.0
#   title 'confirm gcloud version installed'
#   desc 'confirm version reported by gcloud matches the desired version'
#   describe command('gcloud version') do
#     its('stdout') { should include ('Google Cloud SDK 254') }
#   end
# end

control 'terraform version' do
  impact 1.0
  title 'confirm terraform version installed'
  desc 'confirm version reported by terraform matches the desired version'
  describe command('terraform -v') do
    its('stdout') { should include ('0.12') }
  end
end

control 'packer version' do
  impact 1.0
  title 'confirm packer version installed'
  desc 'confirm version reported by packer matches the desired version'
  describe command('packer -v') do
    its('stdout') { should include ('1.4') }
  end
end

control 'chamber version' do
  impact 1.0
  title 'confirm chamber version installed'
  desc 'confirm version reported by chamber matches the desired version'
  describe command('chamber version') do
    its('stdout') { should include ('2.5') }
  end
end

control 'aws-iam-authenticator version' do
  impact 1.0
  title 'confirm aws-iam-authenticator version installed'
  desc 'confirm version reported by aws-iam-authenticator matches the desired version'
  describe command('aws-iam-authenticator') do
    its('exit_status') { should eq 0 }
  end
end

control 'tflint version' do
  impact 1.0
  title 'confirm tflint version installed'
  desc 'confirm version reported by tflint matches the desired version'
  describe command('tflint -v') do
    its('stdout') { should include ('0.11') }
  end
end
